### PR TITLE
[#17] Implement migration entrypoint

### LINKED ISSUE #17: Implement migration
# Clarification and motivation

The administrator should be able to _migrate the DAO_. Currently we propose a simple way to do it: to have a `migrate` entrypoint accepting a single address. This address is supposed to refer to the new contract instance. After a successful `migrate` call the contract enters the "migrated" state permanently, operations can't be performed anymore.

However, this proposed approach has not been approved yet, so we shouldn't start this issue yet.

# Acceptance criteria

A migration approach (maybe not the currently proposed one) is approved, implemented in our code and covered with tests.

### PR DESCRIPTION
**Pull request by sras**
## Description

Updates the specification with the required changes.

## Related issue(s) 

Resolves #17 

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: docs/specification.md [hunk 1]
```diff
@@ -704,9 +707,33 @@ Parameter (in Michelson):
...
+```
+
+Parameter (in Michelson):
+```
+unit
+```
+
+- After a successful `confirm_migration` call, the contract is permanently set
+  to the migrated state, no operations are possible. All contract calls, except
+  `transfer_contract_tokens`, fails with `("MIGRATED", KT1NewDAO)` pair where
```
Comment (gromakovsky) [inline docs/specification.md:732]:
```suggestion
  `transfer_contract_tokens`, fail with `("MIGRATED", KT1NewDAO)` pair where
```

File: src/Lorentz/Contracts/BaseDAO/FA2.hs [hunk 2]
```diff
@@ -12,16 +12,17 @@ module Lorentz.Contracts.BaseDAO.FA2
 import Lorentz
 import Lorentz.Contracts.BaseDAO.Types
 import Lorentz.Contracts.Spec.FA2Interface
+import Lorentz.Contracts.BaseDAO.Management (ensureNotMigrated)
 
 -- | TODO: Implement missing methods
 transfer :: Entrypoint TransferParams Storage
-transfer = doNothing
+transfer = do dip ensureNotMigrated; doNothing
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/FA2.hs:19]: Should be rebased now.

**Comment by gromakovsky** [comment]
Regarding specification, the only missing piece is
> However, it seems that currently we don't do anything with XTZ, so maybe we should just prohibit transferring XTZ to the contract.

So let's write that transferring XTZ to the contract is prohibited because they can't be spent. I think we can implement it in this issue. Maybe it even makes sense to add such a helper to Lorentz and Indigo because it sounds like a common check (it's not the first contract where we want to prohibit it).

Regarding implementation, it looks correct, but lacks tests.

**Comment by sras** [comment]
@gromakovsky Adding some more change. Pls don't review.

**Comment by sras** [comment]
@gromakovsky This is ready for another look now.

File: src/Lorentz/Contracts/BaseDAO/Types.hs [hunk 3]
```diff
@@ -165,3 +165,10 @@ instance CustomErrorHasDoc "nOT_MIGRATION_TARGET" where
   customErrClass = ErrClassActionException
   customErrDocMdCause =
     "Recieved a confirm_migration call on a contract from an address other than the new version"
+
+type instance ErrorArg "xTZ_TRANSFER_IS_FORBIDDEN" = ()
+
+instance CustomErrorHasDoc "xTZ_TRANSFER_IS_FORBIDDEN" where
+  customErrClass = ErrClassActionException
+  customErrDocMdCause =
+    "Recived some XTZ as part of a contract call, which is forbidden"
```
Comment (gromakovsky) [inline src/Lorentz/Contracts/BaseDAO/Types.hs:174]: Received

File: docs/specification.md [hunk 4]
```diff
@@ -202,11 +203,14 @@ The list of erros may be inaccurate and incomplete, it will be updated during th
...
-| `MAX_PROPOSALS_REACHED`   | Throws when trying to propose a proposal when proposals max amount is already reached |
-| `MAX_VOTES_REACHED`   | Throws when trying to vote on a proposal when the votes max amount of that proposal is already reached |
+| `OUT_OF_BOUND_QUORUM_THRESHOLD`| Throws when trying to set quorum threshold that is out of bound from what is specified in the `Config`        |
+| `MAX_PROPOSALS_REACHED`        | Throws when trying to propose a proposal when proposals max amount is already reached |
+| `MAX_VOTES_REACHED`            | Throws when trying to vote on a proposal when the votes max amount of that proposal is already reached |
 | `CONTRACT_MIGRATED`            | Throw when conract has been migrated        |
 | `FROZEN_TOKEN_NOT_TRANSFERABLE`| The entrypoint called does not support frozen tokens         |
+| `MIGRATED`                     | Throw when conract has been migrated        |
+| `NOT_MIGRATING`                | Throw when `confirm_migration` is called and contract is not in migration |
+| `NOT_MIGRATION_TARGET`         | Throw when `confirm_migration` is called by address other than the new address |
```
Comment (gromakovsky) [inline docs/specification.md:213]: `FORBIDDEN_XTZ` should be added here.

**Review by gromakovsky — APPROVED** [review]
Generally LGTM, but please see my comments.

**Pull request merged**